### PR TITLE
feat: Add support for detecting submodules changes

### PR DESCRIPTION
**Pull request by pkit**
Closes: #349

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -2,6 +2,32 @@
...
+  filter="$3"
+  mapfile -t submodules < <(git submodule | awk '{print $2}')
+  for sub in "${submodules[@]}"; do
+    mapfile -t sub_commits < <(
+      git diff "$base" "$sha" -- "$sub" \
+        | grep '^[-+]Subproject commit' \
+        | sed 's/[-+]Subproject commit //'
+      )
+    # this arcane stuff here because of bash<4.4.x idiosyncrasy
+    if [ -n "${sub_commits[0]+"${sub_commits[0]}"}" ]; then
```
Comment (jackton1) [inline entrypoint.sh:17]: Based on https://github.com/tj-actions/changed-files/runs/5290699593?check_suite_focus=true#step:6:91

Line 20: Seems to be generating unbound variable errors

```bash
  D:\a\changed-files\changed-files\.//entrypoint.sh: line 20: sub_commits[1]: unbound variable
```

**Review by jackton1 — CHANGES_REQUESTED** [review]

File: entrypoint.sh [hunk 2]
```diff
@@ -2,6 +2,32 @@
 
 set -eu
 
+function get_diff() {
+  base="$1"
+  sha="$2"
+  filter="$3"
+  mapfile -t submodules < <(git submodule | awk '{print $2}')
```
Comment (jackton1) [inline entrypoint.sh:9]: There also seems to be an issue with using mapfile on Mac which fails silently

See: https://github.com/tj-actions/changed-files/runs/5290697419?check_suite_focus=true#step:3:128

```
All modified files: 
[121](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:121)
  awk: i/o error occurred while closing /dev/stdout
[122](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:122)
   input record number 2, file 
[123](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:123)
   source line number 1
[124](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:124)
  awk: i/o error occurred while closing /dev/stdout
[125](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:125)
   input record number 2, file 
[126](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:126)
   source line number 1
[127](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:127)
  awk: i/o error occurred while closing /dev/stdout
[128](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:128)
   input record number 2, file 
[129](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:129)
   source line number 1
[130](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:130)
  awk: i/o error occurred while closing /dev/stdout
[131](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:131)
   input record number 2, file 
[132](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:132)
   source line number 1
[133](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:133)
awk: i/o error occurred while closing /dev/stdout
[134](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:134)
 input record number 2, file 
[135](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:135)
 source line number 1
[136](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:136)
awk: i/o error occurred while closing /dev/stdout
[137](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:137)
 input record number 2, file 
[138](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:138)
 source line number 1
[139](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:139)
awk: i/o error occurred while closing /dev/stdout
[140](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:140)
 input record number 2, file 
[141](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:141)
 source line number 1
[142](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:142)
awk: i/o error occurred while closing /dev/stdout
[143](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:143)
 input record number 2, file 
[144](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:144)
 source line number 1
[145](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:145)
awk: i/o error occurred while closing /dev/stdout
[146](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:146)
 input record number 2, file 
[147](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:147)
 source line number 1
[148](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:148)
awk: i/o error occurred while closing /dev/stdout
[149](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:149)
 input record number 2, file 
[150](https://github.com/tj-actions/changed-files/runs/5290952730?check_suite_focus=true#step:3:150)
 source line number 1
```

**Review by jackton1 — CHANGES_REQUESTED** [review]

File: entrypoint.sh (hunk 1 above)
Comment (pkit) [reply entrypoint.sh:17]: Yup, here it doesn't understand bash array syntax. Probably old bash

File: entrypoint.sh (hunk 2 above)
Comment (pkit) [reply entrypoint.sh:9]: Looks like I need to target the brain-dead OSX bash. Okay.

File: entrypoint.sh (hunk 1 above)
Comment (pkit) [reply entrypoint.sh]: Removed arrays

**Comment by pkit** [comment]
Also fixed case of "submodule added" or "submodule removed"
If added: all files are added
If removed: nothing checked

**Comment by jackton1** [comment]
@all-contributors Please add @pkit for code

File: entrypoint.sh [hunk 3]
```diff
@@ -2,6 +2,25 @@
 
 set -eu
 
+function get_diff() {
+  base="$1"
+  sha="$2"
```
Comment (jackton1) [inline entrypoint.sh:7]: @pkit This looks good still getting one strange error when a submodule is deleted.

See: https://github.com/tj-actions/changed-files/runs/5296218109?check_suite_focus=true#step:3:89

Screen Shot 2022-02-22 at 6 20 47 PM

**Review by jackton1 — CHANGES_REQUESTED** [review]

File: entrypoint.sh (hunk 3 above)
Comment (pkit) [reply entrypoint.sh:7]: Hmm, will check it out

Comment (jackton1) [reply entrypoint.sh:7]: @pkit Found the issue which was related to missing `submodule: true` which should be include in the checkout action.

Comment (pkit) [reply entrypoint.sh:7]: Yup, seen your commits.

**Pull request merged**